dll/win32/iernonce/dialog.cpp

```diff
@@ -0,0 +1,220 @@
+        SWP_NOMOVE | SWP_NOZORDER | SWP_SHOWWINDOW);
+}
+
+BOOL
+ProgressDlg::ProcessWindowMessage(
+    HWND hwnd, UINT message, WPARAM wParam, LPARAM lParam, LRESULT &lResult,
+    DWORD dwMsgMapID)
+{
+    switch (message)
+    {
+        case WM_INITDIALOG:
+        {
+            if (!m_RunOnceExInst.m_Title.IsEmpty())
+            {
+                SetWindowTextW(m_RunOnceExInst.m_Title);
+            }
+
+            m_hListBox = GetDlgItem(IDC_LB_ITEMS);
+
+            m_hBoldFont = CreateBoldFont(GetFont());
+
+            // Add all sections with non-empty title into listbox
+            int TotalHeight = 0;
+            for (int i = 0; i < m_RunOnceExInst.m_SectionList.GetSize(); i++)
+            {
+                RunOnceExSection &Section = m_RunOnceExInst.m_SectionList[i];
+
+                if (!Section.m_SectionTitle.IsEmpty())
+                {
+                    INT Index = ListBox_AddString(m_hListBox, Section.m_SectionTitle);
+                    TotalHeight += ListBox_GetItemHeight(m_hListBox, Index);
+                    ListBox_SetItemData(m_hListBox, Index, i);
+                }
+            }
+
+            // Remove the WS_EX_CLIENTEDGE exstyle.
+            ::SetWindowLongPtr(m_hListBox, GWL_EXSTYLE, ::GetWindowLongPtr(m_hListBox, GWL_EXSTYLE) & ~WS_EX_CLIENTEDGE);
+
+            ResizeListBoxAndDialog(TotalHeight);
+
+            // Launch a thread.
+            HANDLE hThread = (HANDLE)_beginthreadex(NULL, 0, RunOnceExExecThread, (void *)this, 0, NULL);
+            CloseHandle(hThread);
+
+            break;
+        }
+        case WM_MEASUREITEM:
+        {
+            PMEASUREITEMSTRUCT pMeasureItem = (PMEASUREITEMSTRUCT)lParam;
+
+            RECT TextRect;
+
+            CStringW ItemText;
+            ListBox_GetText(m_hListBox, pMeasureItem->itemID,
+                ItemText.GetBuffer(ListBox_GetTextLen(m_hListBox, pMeasureItem->itemID) + 1));
+
+            CalcTextRect(ItemText, &TextRect);
+
+            ItemText.ReleaseBuffer();
+
+            pMeasureItem->itemHeight = TextRect.bottom - TextRect.top;
+            pMeasureItem->itemWidth  = TextRect.right - TextRect.left;
+
+            break;
+        }
+        case WM_DRAWITEM:
+        {
+            LPDRAWITEMSTRUCT pDrawItem = (PDRAWITEMSTRUCT)lParam;
+            CStringW ItemText;
+
+            ListBox_GetText(m_hListBox, pDrawItem->itemID,
+                            ItemText.GetBuffer(ListBox_GetTextLen(m_hListBox,
+                            pDrawItem->itemID) + 1));
+
+            SetBkMode(pDrawItem->hDC, TRANSPARENT);
+
+            HFONT hOldFont = NULL;
+            if (m_PointedItem == (INT)pDrawItem->itemData)
+            {
+                HDC hCompDC = CreateCompatibleDC(pDrawItem->hDC);
+                HBITMAP hArrowBmp = LoadBitmapW(NULL, MAKEINTRESOURCE(OBM_MNARROW));
+
+                BITMAP bitmap;
+                GetObject(hArrowBmp, sizeof(BITMAP), &bitmap);
```

Comment:
It may be interesting to load this resource only once (+ the bitmap data) in the WM_INITDIALOG handler, store the handle and/or data in a private member of your dialog structure, and then free it in the dialog's WM_DESTROY handler.